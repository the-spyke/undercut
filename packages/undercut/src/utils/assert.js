import { isFunction, isIterable, isIterator } from "./language.js";

export function assert(condition, message) {
	if (!condition) {
		throw new Error(message);
	}
}

export function assertFunctor(functor, name) {
	assert(isFunction(functor), `"${name}" is required, must be a function.`);
}

export function assertPipeline(pipeline) {
	assert(isIterable(pipeline), `"pipeline" is required, must be an iterable of operations.`);
}

export function assertSource(source) {
	assert(isIterable(source) && !isIterator(source), `"source" is required and must be an Iterable, but not an Iterator. It is assumed that a source is re-iterable.`);
}

export function assertSources(sources) {
	sources.forEach(source => assert(isIterable(source), `Every "source" in this operation must be an iterable.`));
}
