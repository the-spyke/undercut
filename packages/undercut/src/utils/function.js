/**
 * @param {any} [value]
 * @returns {any}
 */
export function identity(value) {
	return value;
}

/**
 * Applies boolean negation to the predicate's return value.
 * @param {Function} predicate
 * @returns {Function}
 */
export function negate(predicate) {
	return function (...args) {
		return !predicate(...args);
	};
}

/**
 * Multiplies the predicate's return value by `-1`.
 * @param {Function} predicate
 * @returns {Function}
 */
export function negateSign(operation) {
	return function (...args) {
		return -1 * operation(...args);
	};
}

/**
 * @returns {void}
 */
export function noop() {
	// Empty.
}

/**
 * @param {any} error
 * @returns {void}
 */
export function rethrow(error) {
	throw error;
}
