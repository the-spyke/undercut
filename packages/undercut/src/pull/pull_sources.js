import { assert } from "../utils/assert.js";
import { createIterable } from "../utils/iterable.js";

function* rangeGen(start, end, step) {
	if (start < end) {
		for (let i = start; i < end; i += step) {
			yield i;
		}
	} else if (start > end) {
		for (let i = start; i > end; i -= step) {
			yield i;
		}
	}
}

/**
 * Creates an iterable of numbers in range `[start, end)` with optional step.
 * @param {Number} start
 * @param {Number} end
 * @param {Number} [step=1]
 * @returns {Iterable<Number>}
 */
export function range(start, end, step = 1) {
	assert(Number.isFinite(start), `"start" is required, must be a number.`);
	assert(Number.isFinite(end), `"end" is required, must be a number.`);
	assert(Number.isFinite(step) && step > 0, `"step" must be an absolute non-zero number.`);

	return createIterable(() => rangeGen(start, end, step));
}
