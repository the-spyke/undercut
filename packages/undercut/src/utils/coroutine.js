import { assert } from "./assert.js";
import { getDoneItem, rethrow } from "./function.js";
import { isFunction } from "./language.js";

/**
 * @param {Coroutine} coroutine
 * @param {Error} error
 * @returns {never}
 */
export function abort(coroutine, error) {
	if (isFunction(coroutine.throw)) {
		coroutine.throw(error);
	}

	throw error;
}

/**
 * @type {<T extends Function>(generator: T) => T}
*/
export function asObserver(generator) {
	return function (...args) {
		const genObject = generator(...args);

		genObject.next();

		return genObject;
	};
}

export function asUnclosable(coroutine) {
	return {
		next: coroutine.next.bind(coroutine),
		return: getDoneItem,
		throw: rethrow,
	};
}

/**
 * @type {<T, R>(coroutine: T, tryBeforeClosing?: (coroutine: T) => R) => R)}
 */
export function close(coroutine, tryBeforeClosing) {
	assert(coroutine, `"coroutine" is required.`);

	try {
		if (tryBeforeClosing) {
			return tryBeforeClosing(coroutine);
		}

		return undefined;
	} catch (error) {
		if (isFunction(coroutine.throw)) {
			coroutine.throw(error);
		}

		throw error;
	} finally {
		if (isFunction(coroutine.return)) {
			coroutine.return();
		}
	}
}

export class Cohort {
	static from(...coroutines) {
		return new Cohort(coroutines);
	}

	/**
	 * @param {Array<Coroutine>} coroutines
	 */
	constructor(coroutines) {
		assert(Array.isArray(coroutines), `"coroutines" is required and must be an array of coroutines.`);

		this.coroutines = coroutines;
		this.errors = [];
		this.isClosed = false;
	}

	get hasErrors() {
		return !this.errors.length;
	}

	get isFine() {
		return !this.isClosed && !this.errors.length;
	}

	get length() {
		return this.coroutines.length;
	}

	[Symbol.iterator]() {
		return this.coroutines.values();
	}

	next(coroutine) {
		assert(!this.isClosed, `Trying to add a coroutine to a closed trap.`);

		this.coroutines.push(coroutine);
	}

	return() {
		if (this.isClosed) {
			return;
		}

		this.$close();
	}

	throw(error) {
		if (this.isClosed) {
			throw error;
		}

		this.errors.push(error);
		this.$close();
	}

	$close() {
		const errors = this.errors;

		for (const coroutine of this.coroutines) {
			try {
				if (errors.length && isFunction(coroutine.throw)) {
					coroutine.throw(errors[0]);
				} else if (isFunction(coroutine.return)) {
					coroutine.return();
				}
			} catch (error) {
				if (!errors.length || error !== errors[0]) {
					errors.push(error);
				}
			}
		}

		this.isClosed = true;

		if (errors.length === 1) {
			throw errors[0];
		} else if (errors.length > 1) {
			const error = new Error(`Multiple errors.`);

			error.errors = errors;

			throw error;
		}
	}
}
