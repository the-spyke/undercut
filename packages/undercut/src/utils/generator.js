import { rethrow } from "./function.js";
import { isFunction } from "./language.js";

export function close(closable) {
	if (isFunction(closable.return)) {
		closable.return();
	}
}

function unclosableReturn(value) {
	return { value, done: true };
}

export function asUnclosable(closable) {
	const unclosable = {
		next: closable.next.bind(closable)
	};

	if (isFunction(closable.return)) {
		unclosable.return = unclosableReturn;
	}

	if (isFunction(closable.throw)) {
		unclosable.throw = rethrow;
	}

	return unclosable;
}

export function useClosable(closable, usage) {
	try {
		return usage(closable);
	} finally {
		close(closable);
	}
}
