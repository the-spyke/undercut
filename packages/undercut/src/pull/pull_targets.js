import { assert } from "../utils/assert.js";
import { abort, close } from "../utils/coroutine.js";
import { isObserver } from "../utils/language.js";

/**
 * @param {Iterable} iterable
 * @returns {Array}
 */
export function toArray(iterable) {
	return Array.from(iterable);
}

/**
 * @param {Function} consumer
 * @returns {Function}
 */
export function toConsumer(consumer) {
	return function (iterable) {
		let index = 0;

		for (const item of iterable) {
			consumer(item, index);
			index++;
		}
	};
}

/**
 * @param {Iterable} iterable
 * @returns {Map}
 */
export function toMap(iterable) {
	return new Map(iterable);
}

/**
 * @param {Iterable} iterable
 * @returns {void}
 */
export function toNull(iterable) {
	for (const item of iterable) { // eslint-disable-line no-unused-vars
		// Do nothing.
	}
}

/**
 * @param {Iterable} iterable
 * @returns {Object}
 */
export function toObject(iterable) {
	return Object.fromEntries(iterable);
}

/**
 * @param {Observer} observer
 * @returns {Function}
 */
export function toPushLine(observer) {
	assert(isObserver(observer), `"observer" is required and must be an Observable.`);

	return function (iterable) {
		try {
			for (const item of iterable) {
				observer.next(item);
			}
		} catch (error) {
			abort(observer, error);
		} finally {
			close(observer);
		}

		return observer;
	};
}

/**
 * @param {Iterable} iterable
 * @returns {Set}
 */
export function toSet(iterable) {
	return new Set(iterable);
}

/**
 * @param {Iterable} iterable
 * @returns {any}
 */
export function toValue(iterable) {
	let firstValue = undefined;
	let count = 0;

	for (const item of iterable) {
		assert(count === 0, `"toValue()" may be applied only to a sequence of one item, but got at least 2.`);

		firstValue = item;
		count++;
	}

	assert(count === 1, `"toValue()" may be applied only to a sequence of one item, but got none.`);

	return firstValue;
}
