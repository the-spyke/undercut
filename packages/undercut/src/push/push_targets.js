import { assert, assertFunctor } from "../utils/assert.js";
import { asObserver } from "../utils/coroutine.js";
import { rethrow } from "../utils/function.js";
import { isUndefined, isFunction } from "../utils/language.js";

export function toArray() {
	return {
		next(value) {
			this.values.push(value);
		},
		return() {
			// Empty.
		},
		throw: rethrow,
		values: [],
	};
}

const consumerTarget = asObserver(function* (consumer, finalizer) {
	assertFunctor(consumer, `consumer`);
	assert(isUndefined(finalizer) || isFunction(finalizer), `"finalizer" is optional, could be a function.`);

	let error = undefined;
	let index = 0;

	try {
		while (true) {
			consumer(yield, index);
			index++;
		}
	} catch (e) {
		error = e;
		throw e;
	} finally {
		if (finalizer) {
			finalizer(error, index);
		}
	}
});

/**
 * @param {Function} consumer
 * @param {Function} [finalizer]
 * @returns {Observer}
 */
export function toConsumer(consumer, finalizer) {
	return consumerTarget(consumer, finalizer);
}

/**
 * @returns {Observer}
 */
export function toNull() {
	return {
		next() {
			// Empty.
		},
		return() {
			// Empty.
		},
		throw(e) {
			throw e;
		}
	};
}

const valueTarget = asObserver(function* (setter) {
	assertFunctor(setter, `setter`);

	let firstValue = undefined;
	let count = 0;

	try {
		firstValue = yield;
		count++;
		yield;
		count++;

		assert(count === 1, `"toValue()" may be applied only to a sequence of one item, but got at least 2.`);
	} finally {
		assert(count > 0, `"toValue()" may be applied only to a sequence of one item, but got none.`);

		if (count === 1) {
			setter(firstValue);
		}
	}
});

/**
 * @param {Function} setter
 * @returns {Observer}
 */
export function toValue(setter) {
	return valueTarget(setter);
}
