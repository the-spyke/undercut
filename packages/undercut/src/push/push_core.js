import { assert, assertPipeline, assertSource } from "../utils/assert.js";
import { isObserver, isFunction } from "../utils/language.js";
import { closeObserver, initializeObserver } from "../utils/observer.js";

import { toArray } from "./push_targets.js";

const operationErrorMessage = `An operation must be a function taking and returning an Observer.`;

function connectPipeline(pipeline, target, initialize = true) {
	assertPipeline(pipeline);
	assert(isObserver(target), `"target" is required and must be an Observer.`);

	const operations = Array.isArray(pipeline) ? pipeline : [...pipeline];

	let observer = target;

	for (let index = operations.length - 1; index >= 0; index--) {
		const operation = operations[index];

		assert(isFunction(operation), operationErrorMessage);

		observer = operation(observer);

		assert(isObserver(observer), operationErrorMessage);

		if (index > 0 || initialize) {
			initializeObserver(observer);
		}
	}

	return observer;
}

export function composeOperations(operations) {
	return function (observer) {
		return connectPipeline(operations, observer, false);
	};
}

export function pushLine(pipeline, target) {
	return connectPipeline(pipeline, target, true);
}

export function push(target, pipeline, source) {
	assertSource(source);

	const observer = pushLine(pipeline, target);

	try {
		for (const item of source) {
			observer.next(item);
		}
	} finally {
		closeObserver(observer);
	}

	return target;
}

export function pushItems(pipeline, source) {
	const target = toArray();

	push(target, pipeline, source);

	return target.values;
}
