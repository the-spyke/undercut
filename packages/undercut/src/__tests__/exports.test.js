import * as pullExports from "../../pull.js";
import * as pushExports from "../../push.js";
import * as utilsExports from "../../utils.js";

describe(`exports`, () => {
	test(`pull`, () => {
		expect(Object.keys(pullExports).sort()).toMatchSnapshot();
	});

	test(`push`, () => {
		expect(Object.keys(pushExports).sort()).toMatchSnapshot();
	});

	test(`utils`, () => {
		expect(Object.keys(utilsExports).sort()).toMatchSnapshot();
	});
});
