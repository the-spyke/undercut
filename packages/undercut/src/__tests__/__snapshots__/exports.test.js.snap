// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports pull 1`] = `
Array [
  "append",
  "asc",
  "average",
  "chunk",
  "closeIterator",
  "compact",
  "compare",
  "composeOperations",
  "concatEnd",
  "concatStart",
  "count",
  "desc",
  "difference",
  "differenceBy",
  "every",
  "filter",
  "find",
  "findIndex",
  "first",
  "flatten",
  "flattenIterables",
  "forEach",
  "getIterator",
  "groupBy",
  "includes",
  "interleave",
  "intersection",
  "intersectionBy",
  "join",
  "last",
  "makeReiterable",
  "map",
  "max",
  "min",
  "nth",
  "orderBy",
  "prepend",
  "pull",
  "pullItems",
  "pullLine",
  "range",
  "reduce",
  "remove",
  "reverse",
  "skip",
  "skipWhile",
  "some",
  "sort",
  "sortNumbers",
  "sortStrings",
  "sum",
  "symmetricDifference",
  "symmetricDifferenceBy",
  "take",
  "takeWhile",
  "toArray",
  "toConsumer",
  "toMap",
  "toNull",
  "toObject",
  "toPushLine",
  "toSet",
  "toValue",
  "union",
  "unionBy",
  "unique",
  "uniqueBy",
  "unzip",
  "unzipWith",
  "useIterator",
  "zip",
  "zipWith",
]
`;

exports[`exports push 1`] = `
Array [
  "append",
  "asc",
  "average",
  "chunk",
  "closeObserver",
  "compact",
  "compare",
  "composeOperations",
  "concatEnd",
  "concatStart",
  "count",
  "desc",
  "difference",
  "differenceBy",
  "every",
  "filter",
  "find",
  "findIndex",
  "first",
  "flatten",
  "flattenIterables",
  "forEach",
  "groupBy",
  "includes",
  "initializeObserver",
  "interleave",
  "intersection",
  "intersectionBy",
  "join",
  "last",
  "map",
  "max",
  "min",
  "nth",
  "orderBy",
  "prepend",
  "push",
  "pushItems",
  "pushLine",
  "reduce",
  "remove",
  "reverse",
  "skip",
  "skipWhile",
  "some",
  "sort",
  "sortNumbers",
  "sortStrings",
  "sum",
  "symmetricDifference",
  "symmetricDifferenceBy",
  "take",
  "takeWhile",
  "toArray",
  "toConsumer",
  "toNull",
  "toValue",
  "union",
  "unionBy",
  "unique",
  "uniqueBy",
  "unzip",
  "unzipWith",
  "useObserver",
  "zip",
  "zipWith",
]
`;

exports[`exports utils 1`] = `
Array [
  "asc",
  "closeIterator",
  "closeObserver",
  "compare",
  "delay",
  "desc",
  "getIterator",
  "identity",
  "initializeObserver",
  "isBoolean",
  "isDefined",
  "isFalsy",
  "isFunction",
  "isIterable",
  "isIterator",
  "isNegative",
  "isNegativeOrZero",
  "isNull",
  "isNullish",
  "isNumber",
  "isNumberValue",
  "isObject",
  "isObjectValue",
  "isObserver",
  "isPositive",
  "isPositiveOrZero",
  "isString",
  "isSymbol",
  "isTruthy",
  "isUndefined",
  "makeReiterable",
  "negate",
  "negateSign",
  "noop",
  "rethrow",
  "unwrapPromise",
  "useIterator",
  "useObserver",
  "wait",
]
`;
